File: mns-reactor/src/main/java/cz/fb/manaus/reactor/profit/MatchbookProfitPlugin.java
package cz.fb.manaus.reactor.profit;

import com.google.common.collect.ImmutableMap;
import cz.fb.manaus.core.model.SettledBet;
import org.springframework.context.annotation.Profile;
import org.springframework.stereotype.Component;

import java.util.List;
import java.util.Map;

@Component
@Profile("matchbook")
public class MatchbookProfitPlugin implements ProfitPlugin {

    @Override
    public Map<String, Double> getCharges(List<SettledBet> bets, double chargeRate) {
        var result = ImmutableMap.<String, Double>builder();
        for (var bet : bets) {
            var betId = bet.getBetAction().getBetId();
            result.put(betId, getCharge(chargeRate, bet.getProfitAndLoss(), bet.getPrice().getAmount()));
        }
        return result.build();
    }

    double getCharge(double chargeRate, double profitAndLoss, double amount) {
        if (profitAndLoss < 0) {
            return Math.min(amount, -profitAndLoss) * chargeRate;
        } else {
            return profitAndLoss * chargeRate;
        }
    }
}

File: mns-reactor/src/main/java/cz/fb/manaus/reactor/profit/MatchbookProfitPlugin.kt
package cz.fb.manaus.reactor.profit

import cz.fb.manaus.core.model.RealizedBet
import cz.fb.manaus.core.model.SettledBet
import org.springframework.context.annotation.Profile
import org.springframework.stereotype.Component
import kotlin.math.min

@Component
@Profile("matchbook")
class MatchbookProfitPlugin : ProfitPlugin {

    override fun getCharges(bets: List<RealizedBet>, chargeRate: Double): Map<String, Double> {
        return bets.asSequence().map { it.settledBet }
                .map { it.id to getCharge(chargeRate, it) }
                .toMap()
    }

    private fun getCharge(chargeRate: Double, bet: SettledBet): Double {
        return if (bet.commission != null) {
            bet.commission!!
        } else {
            getCountedCharge(chargeRate, bet.profitAndLoss, bet.price.amount)
        }
    }

    internal fun getCountedCharge(chargeRate: Double, profitAndLoss: Double, amount: Double): Double {
        return if (profitAndLoss < 0) {
            min(amount, -profitAndLoss) * chargeRate
        } else {
            profitAndLoss * chargeRate
        }
    }
}
